expat/tests/basic_tests.c

```diff
@@ -5220,84 +5214,68 @@ START_TEST(test_big_tokens_take_linear_time) {
-  // (2) small enough that the test actually catches regressions.
-  const int max_slowdown = 15;
   char aaaaaa[4096];
   const int fillsize = (int)sizeof(aaaaaa);
   const int fillcount = 100;
+  const unsigned approx_bytes = fillsize * fillcount; // ignore pre/post.
+  const unsigned max_factor = 4;
+  const unsigned max_scanned = max_factor * approx_bytes;
 
   memset(aaaaaa, 'a', fillsize);
 
   if (! g_reparseDeferralEnabledDefault) {
     return; // heuristic is disabled; we would get O(n^2) and fail.
   }
-#if ! defined(__linux__)
-  if (CLOCKS_PER_SEC < 100000) {
-    // Skip this test if clock() doesn't have reasonably good resolution.
-    // This workaround is primarily targeting Windows and FreeBSD, since
-    // XSI requires the value to be 1.000.000 (10x the condition here), and
-    // we want to be very sure that at least one platform in CI can catch
-    // regressions (through a failing test).
-    return;
-  }
-#endif
 
-  clock_t baseline = 0;
   for (int i = 0; i < num_cases; ++i) {
     XML_Parser parser = XML_ParserCreate(NULL);
     assert_true(parser != NULL);
     enum XML_Status status;
-    set_subtest("max_slowdown=%d text=\"%saaaaaa%s\"", max_slowdown,
-                text[i].pre, text[i].post);
-    const clock_t start = clock();
+    set_subtest("text=\"%saaaaaa%s\"", text[i].pre, text[i].post);
 
     // parse the start text
+    g_bytesScanned = 0;
     status = _XML_Parse_SINGLE_BYTES(parser, text[i].pre,
                                      (int)strlen(text[i].pre), XML_FALSE);
     if (status != XML_STATUS_OK) {
       xml_failure(parser);
     }
+
     // parse lots of 'a', failing the test early if it takes too long
+    unsigned past_max_count = 0;
     for (int f = 0; f < fillcount; ++f) {
       status = _XML_Parse_SINGLE_BYTES(parser, aaaaaa, fillsize, XML_FALSE);
       if (status != XML_STATUS_OK) {
         xml_failure(parser);
       }
-      // i == 0 means we're still calculating the baseline value
-      if (i > 0) {
-        const clock_t now = clock();
-        const clock_t clocks_so_far = now - start;
-        const int slowdown = clocks_so_far / baseline;
-        if (slowdown >= max_slowdown) {
-          fprintf(
-              stderr,
-              "fill#%d: clocks_so_far=%d baseline=%d slowdown=%d max_slowdown=%d\n",
-              f, (int)clocks_so_far, (int)baseline, slowdown, max_slowdown);
-          fail(too_slow_failure_message);
-        }
+      if (g_bytesScanned > max_scanned) {
+        const unsigned pushed
+            = (unsigned)strlen(text[i].pre) + (f + 1) * fillsize;
+        fprintf(
+            stderr,
+            "after %d/%d loops: pushed=%u scanned=%u (factor ~%.2f) max_scanned: %u (factor ~%u)\n",
+            f + 1, fillcount, pushed, g_bytesScanned,
+            g_bytesScanned / (double)pushed, max_scanned, max_factor);
+        past_max_count++; // used to allow a few logs, so the trend is clear
+        assert_true(past_max_count < 5);
```

Comment:
- I'm not sure what a "log" would be in this context, and I don#t yet get the idea of this counter. Please help me understand.
- The error message could be printed if `g_bytesScanned > max_scanned` and `past_max_count < 4`. I'd expect that we want a test that is all silent unless there is a test failure, and at least superficially this block of code seems to violate that approach. I did not see the error message in practice, but it would be nice if I am the last person to wonder about this.

Reply:
I want to get a few of those "after x/y loops" logs out to make the quadratic behavior clearer on failure. Try removing the early `return` on line 5227 for an example.

If we ever enter this block, the test _will_ fail: either after 5 prints of that log, or by the same check on line 5272. Technically, the 5-times rule isn't needed for correctness; only for making the test fail within a reasonable amount of time. Again, try removing the early `return` along with this early failure and see how long you have to wait -- I'll be very impressed if you have the patience to let the test finish. ;)

Reply:
@Snild-Sony thanks for elaborating, makes more sense now. I think part of my earlier confusion was a language thing because "log" to me is either a noun for a whole log or a verb, while you use "log" here for "log entries" or for a nounified verb in plural form… Maybe "loggings" (as in actions) might have been clear to me.

So the added complexity is to only produce 5 lines of logging output rather than 100 per chunk size and to exit early on sure soon failure, I see.

My vote for adding a comment that `assert_true(past_max_count < 5);` is about exiting the test early rather than about test correctness: it is not a "classic test assertion", right? Maybe it should not even be an assert then but `if (xxxxxx) fail(...)` with a self-explaining error message.